### PR TITLE
tcp api plans

### PR DESCRIPTION
**Pull request by Okhelifi**
## Issue

https://gravitee.atlassian.net/browse/APIM-3589

## Description

Filter plans available when user is inAPI creation workflow and use TCP listener.
Update REST API to allow user to create an API only with a keyless plan

## Additional context

---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-kuiwvwlpsl.chromatic.com)

### CODE REVIEW

File: gravitee-apim-console-webui/src/services-ngx/constants.service.ts [hunk 1]
```diff
@@ -72,4 +72,22 @@ export class ConstantsService {
       return matchedSecurityType.length > 0 && matchedSecurityType[0].enabled;
     });
   }
+
+  getPlanMenuItems(definitionVersion: DefinitionVersion, listenerTypes: ListenerType[]): PlanMenuItemVM[] {
+    const availablePlanMenuItems = this.getEnabledPlanMenuItems();
+
+    if (definitionVersion === 'V4' && listenerTypes && listenerTypes.every((listenerType) => listenerType === 'TCP')) {
```
Comment (ytvnr) [inline gravitee-apim-console-webui/src/services-ngx/constants.service.ts]:
```suggestion
    if (definitionVersion === 'V4' && listenerTypes?.every((listenerType) => listenerType === 'TCP')) {
```

File: gravitee-apim-console-webui/src/services-ngx/constants.service.ts [hunk 2]
```diff
@@ -72,4 +72,22 @@ export class ConstantsService {
...
+      return availablePlanMenuItems.filter((p) => p.planFormType === 'KEY_LESS');
+    }
+
+    if (definitionVersion === 'V4' && listenerTypes && listenerTypes.every((listenerType) => listenerType === 'SUBSCRIPTION')) {
```
Comment (ytvnr) [inline gravitee-apim-console-webui/src/services-ngx/constants.service.ts]: same

File: gravitee-apim-console-webui/src/services-ngx/constants.service.ts [hunk 3]
```diff
@@ -72,4 +72,22 @@ export class ConstantsService {
...
+      return availablePlanMenuItems.filter((planMenuItem) => planMenuItem.planFormType === 'PUSH');
+    }
+
+    if (definitionVersion !== 'V4' || (listenerTypes && listenerTypes.every((listenerType) => ['HTTP', 'TCP'].includes(listenerType)))) {
```
Comment (ytvnr) [inline gravitee-apim-console-webui/src/services-ngx/constants.service.ts]: same

File: gravitee-apim-console-webui/src/services-ngx/constants.service.ts (hunk 1 above)
Comment (ytvnr) [inline gravitee-apim-console-webui/src/services-ngx/constants.service.ts]: it is not a blocker, but those "if" need a bit of effort to understand, we can maybe extract them in meaningful private method ?

File: gravitee-apim-e2e/api-test/src/apis/plans/mapi-v2/create-tcp-api-plan.spec.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        envId,
+        apiId: importedApi.id,
+        createPlan: MAPIV2PlansFaker.newPlanV4(),
+      }),
+    );
+
+    expect(createdPlan).toBeTruthy();
+  });
+
+  test('should not be able to add a api-key plan', async () => {
```
Comment (ytvnr) [inline gravitee-apim-e2e/api-test/src/apis/plans/mapi-v2/create-tcp-api-plan.spec.ts]: `it.each''` to test all the other plans ?
As you are using inclusion and exclusion in the implementation, the business rule can break if we add a new plan

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/apim/core/plan/domain_service/CreatePlanDomainServiceTest.java [hunk 5]
```diff
@@ -89,6 +94,7 @@ class CreatePlanDomainServiceTest {
     private static final String TAG = "tag1";
 
     private static final Api API_PROXY_V4 = aProxyApiV4().toBuilder().id(API_ID).build();
+    private static final Api TCP_PROXY_API_V4 = aTcpApiV4().toBuilder().id(API_ID).build();
```
Comment (ytvnr) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/apim/core/plan/domain_service/CreatePlanDomainServiceTest.java:96]: maybe you can rename the previous line with HTTP_PROXY_V4

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/apim/core/plan/use_case/CreatePlanUseCaseTest.java [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+    void setUp() {
+        apiCrudService.initWith(List.of(API));
+        parametersQueryService.initWith(
+            List.of(new Parameter(Key.PLAN_SECURITY_KEYLESS_ENABLED.key(), ENVIRONMENT_ID, ParameterReferenceType.ENVIRONMENT, "true"))
+        );
+    }
+
+    @AfterEach
+    void tearDown() {
+        apiCrudService.reset();
```
Comment (ytvnr) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/apim/core/plan/use_case/CreatePlanUseCaseTest.java:121]: why not reseting `parameterQueryService` ?

**Review by ytvnr — COMMENTED** [review]
good job

**Review by ytvnr — APPROVED** [review]

**Comment by Okhelifi** [comment]
> ## 2 failed tests on [run #501 ↗︎](https://cloud.cypress.io/projects/88vnc3/runs/501/overview?utm_source=github&utm_medium=failed&utm_campaign=run%20number)
> [](https://cloud.cypress.io/projects/88vnc3/runs/501/overview?reviewViewBy=FAILED&utm_source=github&utm_medium=failed&utm_campaign=failed%20tests) [2](https://cloud.cypress.io/projects/88vnc3/runs/501/overview?reviewViewBy=FAILED&utm_source=github&utm_medium=failed&utm_campaign=failed%20tests)	[](https://cloud.cypress.io/projects/88vnc3/runs/501/overview?utm_source=github&utm_medium=failed&utm_campaign=passed%20tests) [73](https://cloud.cypress.io/projects/88vnc3/runs/501/overview?utm_source=github&utm_medium=failed&utm_campaign=passed%20tests)	[](https://cloud.cypress.io/projects/88vnc3/runs/501/test-results?utm_source=github&utm_medium=failed&utm_campaign=pending%20tests&statuses=%5B%7B%22value%22%3A%22PENDING%22%2C%22label%22%3A%22PENDING%22%7D%5D) [0](https://cloud.cypress.io/projects/88vnc3/runs/501/test-results?utm_source=github&utm_medium=failed&utm_campaign=pending%20tests&statuses=%5B%7B%22value%22%3A%22PENDING%22%2C%22label%22%3A%22PENDING%22%7D%5D)	[](https://cloud.cypress.io/projects/88vnc3/runs/501/test-results?utm_source=github&utm_medium=failed&utm_campaign=skipped%20tests&statuses=%5B%7B%22value%22%3A%22SKIPPED%22%2C%22label%22%3A%22SKIPPED%22%7D%5D) [0](https://cloud.cypress.io/projects/88vnc3/runs/501/test-results?utm_source=github&utm_medium=failed&utm_campaign=skipped%20tests&statuses=%5B%7B%22value%22%3A%22SKIPPED%22%2C%22label%22%3A%22SKIPPED%22%7D%5D)	[Flakiness](https://cloud.cypress.io/projects/88vnc3/runs/501/overview?reviewViewBy=FLAKY&utm_source=github&utm_medium=failed&utm_campaign=flaky%20tests) [0](https://cloud.cypress.io/projects/88vnc3/runs/501/overview?reviewViewBy=FLAKY&utm_source=github&utm_medium=failed&utm_campaign=flaky%20tests)
> Details:
> 
> feat: update REST API to allow keyless plan on TCP API
> Project: APIM UI Tests	Commit: `050b0b523a`
> Status: Failed	Duration: 05:54 [💡](https://on.cypress.io/parallelization?utm_source=github&utm_medium=failed&utm_campaign=parallelization%20docs)
> Started: Jan 18, 2024 4:41 PM	Ended: Jan 18, 2024 4:47 PM
> Failed  ui-test/integration/apim/ui/apis/ui-api-plans.spec.ts • 2 failed tests
> [`View`](https://cloud.cypress.io/projects/88vnc3/runs/501/test-results?utm_source=github&utm_medium=failed&utm_campaign=view%20spec&specs=%5B%7B%22value%22%3A%22%5B%5C%22bb97d7d8-7ff0-4059-8925-a26a0d88270c%5C%22%5D%22%2C%22label%22%3A%22ui-test%2Fintegration%2Fapim%2Fui%2Fapis%2Fui-api-plans.spec.ts%22%7D%5D) [`Output`](https://cloud.cypress.io/projects/88vnc3/runs/501/test-results/instance/bb97d7d8-7ff0-4059-8925-a26a0d88270c/stdout?utm_source=github&utm_medium=failed&utm_campaign=view%20spec%20stdout) [`Video`](https://cloud.cypress.io/projects/88vnc3/runs/501/specs/bb97d7d8-7ff0-4059-8925-a26a0d88270c/video?utm_source=github&utm_medium=failed&utm_campaign=view%20spec%20video)
> 
> Test	Artifacts
> [ API Plans Feature > Create a New Plan (Keyless), publish the plan and delete plan ](https://cloud.cypress.io/projects/88vnc3/runs/501/overview/606fc354-bf68-400b-9bd8-699b9b6921ef?reviewViewBy=FAILED&utm_source=github&utm_medium=failed&utm_campaign=view%20test)	[ `Test Replay` ](https://cloud.cypress.io/projects/88vnc3/runs/501/test-results/606fc354-bf68-400b-9bd8-699b9b6921ef/replay?utm_source=github&utm_medium=failed&utm_campaign=view%20test%20replay) [ `Screenshots` ](https://cloud.cypress.io/projects/88vnc3/runs/501/overview/606fc354-bf68-400b-9bd8-699b9b6921ef/screenshots?reviewViewBy=FAILED&utm_source=github&utm_medium=failed&utm_campaign=view%20test%20screenshots) [ `Video` ](https://cloud.cypress.io/projects/88vnc3/runs/501/test-results/606fc354-bf68-400b-9bd8-699b9b6921ef/video?utm_source=github&utm_medium=failed&utm_campaign=view%20test%20video)
> [ API Plans Feature > Create a New Plan (Keyless), publish then deprecate the plan and delete plan ](https://cloud.cypress.io/projects/88vnc3/runs/501/overview/b85f6df9-f67d-435e-940d-dce9c415b633?reviewViewBy=FAILED&utm_source=github&utm_medium=failed&utm_campaign=view%20test)	[ `Test Replay` ](https://cloud.cypress.io/projects/88vnc3/runs/501/test-results/b85f6df9-f67d-435e-940d-dce9c415b633/replay?utm_source=github&utm_medium=failed&utm_campaign=view%20test%20replay) [ `Screenshots` ](https://cloud.cypress.io/projects/88vnc3/runs/501/overview/b85f6df9-f67d-435e-940d-dce9c415b633/screenshots?reviewViewBy=FAILED&utm_source=github&utm_medium=failed&utm_campaign=view%20test%20screenshots) [ `Video` ](https://cloud.cypress.io/projects/88vnc3/runs/501/test-results/b85f6df9-f67d-435e-940d-dce9c415b633/video?utm_source=github&utm_medium=failed&utm_campaign=view%20test%20video)
> ### Review all test suite changes for [PR #6345 ↗︎](https://cloud.cypress.io/projects/88vnc3/branches/apim-3589-tcp-api-plans-run-e2e/review?utm_source=github&utm_medium=failed&utm_campaign=review%20changes&changeRequest=6345)

Just tried locally and everything it's ok 

Capture d’écran 2024-01-18 à 17 58 15

**Pull request merged**